### PR TITLE
[docs] Adds a doc on LoRA support for diffusers

### LINKED ISSUE #2028: Write a doc for centralizing our LoRA efforts
Currently, we have incoming support for LoRA to fine-tune with [DreamBooth](https://github.com/huggingface/diffusers/pull/1884) and [Text2Image](https://github.com/huggingface/diffusers/pull/2002). 

These things enable the community to run fine-tuning on consumer GPUs which is a big enabling factor for many use cases. 

So with that, I think it's important to put together a doc highlighting these efforts. This doc could live under https://github.com/huggingface/diffusers/tree/main/docs/source/en/training (as suggested by @patil-suraj). 

I will start working on this doc next week. If anyone has any suggestions, feel free to comment. Opening this issue thread so that things don't get slipped through the cracks.

### PR DESCRIPTION
**Pull request by sayakpaul**
Closes #2028.

### CODE REVIEW

File: examples/text_to_image/README.md [hunk 1]
```diff
@@ -162,9 +162,9 @@ accelerate --mixed_precision="fp16" launch train_text_to_image_lora.py \
 
 The above command will also run inference as fine-tuning progresses and log the results to Weights and Biases.
 
-**___Note: When using LoRA we can use a much higher learning rate compared to non-LoRA fine-tuning. Here we use *1e-4* instead of the usual *1e-5*. Also, by using LoRA, it's possible to run `train_text_to_image_lora.py` in consumer GPUs like T4 or V100.**
```
Comment (sayakpaul) [inline examples/text_to_image/README.md:165]: I know this is not the characteristic of an ideal PR to make changes to unrelated files. But these seemed pretty minor.

File: docs/source/en/training/lora.mdx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+Assuming, you used the `examples/text_to_image/train_text_to_image_lora.py` to fine-tune Stable Diffusion on the [Pokemons
+dataset](https://huggingface.co/lambdalabs/pokemon-blip-captions), you can perform inference like so: 
+
+```py 
+from diffusers import StableDiffusionPipeline
+import torch
+
+model_path = "sayakpaul/sd-model-finetuned-lora-t4"
+pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", torch_dtype=torch.float16)
```
Comment (patrickvonplaten) [inline docs/source/en/training/lora.mdx:110]: (nit) maybe we can show how to retrieve the base_model from the model card by loading the yaml code via `huggingface_hub`

File: examples/text_to_image/README.md (hunk 1 above)
Comment (patrickvonplaten) [reply examples/text_to_image/README.md:165]: Thanks for correcting it :-)

File: docs/source/en/training/lora.mdx (hunk 2 above)
Comment (patrickvonplaten) [reply docs/source/en/training/lora.mdx:110]: https://github.com/huggingface/huggingface_hub/blob/5a12851f54bf614be39614034ed3a9031922d297/src/huggingface_hub/repocard.py#L136

File: docs/source/en/training/lora.mdx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    * Number of training steps
+* Inference time 
+    * Number of steps 
+    * Scheduler type
+
+Additionally, you can follow [this blog](https://huggingface.co/blog/dreambooth) that documents some of our experimental
+findings for performing DreamBooth training Stable Diffusion.
+
+When fine-tuning, the LoRA update matrices are only added to the attention layers. To enable this, we added new weight
+loading functionalities. Their details are available [here](https://huggingface.co/docs/diffusers/main/en/api/loaders).
```
Comment (pcuenca) [inline docs/source/en/training/lora.mdx:98]: Is this because of this particular implementation or something general in LoRA? (Just curious, no need to change anything here).

File: docs/source/en/training/lora.mdx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+## Inference 
+
+Assuming, you used the `examples/text_to_image/train_text_to_image_lora.py` to fine-tune Stable Diffusion on the [Pokemons
```
Comment (pcuenca) [inline docs/source/en/training/lora.mdx:102]:
```suggestion
Assuming you used the `examples/text_to_image/train_text_to_image_lora.py` to fine-tune Stable Diffusion on the [Pokemon
```

File: docs/source/en/training/lora.mdx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+dataset](https://huggingface.co/lambdalabs/pokemon-blip-captions), you can perform inference like so:
```
Comment (pcuenca) [inline docs/source/en/training/lora.mdx:103]:
```suggestion
dataset](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions), you can perform inference like so: 
```

File: docs/source/en/training/lora.mdx (hunk 4 above)
Comment (pcuenca) [reply docs/source/en/training/lora.mdx:102]: Also, the previous example was about dreambooth, not fine-tuning. Maybe include a brief mention before this section? Something like `A similar process can be followed to fully fine-tune Stable Diffusion on a custom dataset using the train_text_to_image_lora script`

File: docs/source/en/training/lora.mdx (hunk 2 above)
Comment (pcuenca) [reply docs/source/en/training/lora.mdx:110]: Nice! Maybe include it as a tip below the current snippet?

File: docs/source/en/training/lora.mdx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import torch
+
+model_path = "sayakpaul/sd-model-finetuned-lora-t4"
+pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", torch_dtype=torch.float16)
+pipe.unet.load_attn_procs(model_path)
+pipe.to("cuda")
+
+prompt = "A pokemon with green eyes and red legs."
+image = pipe(prompt, num_inference_steps=30, guidance_scale=7.5).images[0]
+image.save("pokemon.png")
```
Comment (pcuenca) [inline docs/source/en/training/lora.mdx:116]: Just wondering, maybe display the image here? We never do it in the docs, what's your opinion about starting doing it to make things more visual?

File: docs/source/en/training/lora.mdx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+```
+
+[`sayakpaul/sd-model-finetuned-lora-t4`](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4) contains [LoRA fine-tuned update matrices](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4/blob/main/pytorch_lora_weights.bin)
+which is only 3 MBs in size. During inference, the pre-trained Stable Diffusion checkpoints loaded alongside these update
```
Comment (pcuenca) [inline docs/source/en/training/lora.mdx:120]:
```suggestion
which is only 3 MBs in size. During inference, the pre-trained Stable Diffusion checkpoints are loaded alongside these update
```

File: examples/text_to_image/README.md [hunk 8]
```diff
@@ -191,7 +191,7 @@ image.save("pokemon.png")
 
 For faster training on TPUs and GPUs you can leverage the flax training example. Follow the instructions above to get the model and dataset before running the script.
 
-____Note: The flax example don't yet support features like gradient checkpoint, gradient accumulation etc, so to use flax for faster training we will need >30GB cards.___
+**___Note: The flax example don't yet support features like gradient checkpoint, gradient accumulation etc, so to use flax for faster training we will need >30GB cards.___**
```
Comment (pcuenca) [inline examples/text_to_image/README.md:194]:
```suggestion
**___Note: The flax example doesn't yet support features like gradient checkpoint, gradient accumulation etc, so to use flax for faster training we will need >30GB cards or TPU v3.___**
```

(I'm assuming it doesn't work on TPU v2, but never tried).

**Review by pcuenca — APPROVED** [review]
Awesome, thanks!

File: docs/source/en/training/lora.mdx (hunk 4 above)
Comment (patil-suraj) [reply docs/source/en/training/lora.mdx:102]: +1

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for working on this. Looks great!

File: docs/source/en/training/lora.mdx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+- Previous pretrained weights are kept frozen so that model is not prone to [catastrophic forgetting](https://www.pnas.org/doi/10.1073/pnas.1611835114). 
+- Rank-decomposition matrices have significantly fewer parameters than original model, which means that trained LoRA weights are easily portable.
+- LoRA attention layers allow to control to which extent the model is adapted toward new training images via a `scale` parameter.
+
+[cloneofsimo](https://github.com/cloneofsimo) was the first to try out LoRA training for Stable Diffusion in the popular [lora](https://github.com/cloneofsimo/lora) GitHub repository.
+
+<Tip>
+
+LoRA also allows us to achieve greater memory efficiency since the pretrained weights are kept frozen, only the LoRA weights are trained, thereby
+allowing us to run fine-tuning on consumer GPUs like Tesla T4.
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply docs/source/en/training/lora.mdx:32]: Update: text_to_image finetuning works on a 2080 Ti!

```suggestion
allowing us to run fine-tuning on consumer GPUs like Tesla T4, RTX 3080 or even RTX 2080 Ti!
```

File: docs/source/en/training/lora.mdx (hunk 2 above)
Comment (patrickvonplaten) [reply docs/source/en/training/lora.mdx:110]: For me it's fine in the same code snippet

File: docs/source/en/training/lora.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/training/lora.mdx:32]: Goodness!

File: docs/source/en/training/lora.mdx (hunk 3 above)
Comment (sayakpaul) [reply docs/source/en/training/lora.mdx:98]: I think the bit on only considering attention layers for adding LoRA weights is coming from the original LoRA paper. There, the authors only experimented with the attention layers.

File: docs/source/en/training/lora.mdx (hunk 6 above)
Comment (sayakpaul) [reply docs/source/en/training/lora.mdx:116]: Diffusion for computer vision is definitely about visuals. I like the idea and I think we should definitely add it :)

File: docs/source/en/training/lora.mdx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+Low-Rank Adaption of Large Language Models was first introduced by Microsoft in
+[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) by *Edward J. Hu, Yelong Shen, Phillip Wallis, Zeyuan Allen-Zhu, Yuanzhi Li, Shean Wang, Lu Wang, Weizhu Chen*.
+
+In a nutshell, LoRA allows adapting pretrained models by adding pairs of rank-decomposition weight matrices (called **update marrices**) 
+to existing weights and **only** training those newly added weights. This has a couple of advantages:
+
+- Previous pretrained weights are kept frozen so that model is not prone to [catastrophic forgetting](https://www.pnas.org/doi/10.1073/pnas.1611835114). 
+- Rank-decomposition matrices have significantly fewer parameters than original model, which means that trained LoRA weights are easily portable.
+- LoRA attention layers allow to control to which extent the model is adapted toward new training images via a `scale` parameter.
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply docs/source/en/training/lora.mdx:25]: See if the current edits make sense.

File: docs/source/en/training/lora.mdx (hunk 6 above)
Comment (sayakpaul) [reply docs/source/en/training/lora.mdx:116]: Added an image.

File: docs/source/en/training/lora.mdx [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+Low-Rank Adaption of Large Language Models was first introduced by Microsoft in
+[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) by *Edward J. Hu, Yelong Shen, Phillip Wallis, Zeyuan Allen-Zhu, Yuanzhi Li, Shean Wang, Lu Wang, Weizhu Chen*.
+
+In a nutshell, LoRA allows adapting pretrained models by adding pairs of rank-decomposition weight matrices (called **update matrices**) 
+to existing weights and **only** training those newly added weights. This has a couple of advantages:
+
+- Previous pretrained weights are kept frozen so that the model is not so prone to [catastrophic forgetting](https://www.pnas.org/doi/10.1073/pnas.1611835114). 
+- Rank-decomposition matrices have significantly fewer parameters than the original model, which means that trained LoRA weights are easily portable.
+- LoRA matrices are generally added to the attention layers of the original model and they control to control to which extent the model is adapted toward new training images via a `scale` parameter.
```
Comment (pcuenca) [inline docs/source/en/training/lora.mdx:25]:
```suggestion
- LoRA matrices are generally added to the attention layers of the original model and they control to which extent the model is adapted toward new training images via a `scale` parameter.
```

File: docs/source/en/training/lora.mdx [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+**__Note that the usage of LoRA is not limited to only attention layers. In the original LoRA work, the authors found out that just ammending
```
Comment (pcuenca) [inline docs/source/en/training/lora.mdx:27]:
```suggestion
**__Note that the usage of LoRA is not just limited to attention layers. In the original LoRA work, the authors found out that just amending
```

**Review by pcuenca — APPROVED** [review]
Awesome, thanks a lot!

File: docs/source/en/training/lora.mdx (hunk 2 above)
Comment (sayakpaul) [reply docs/source/en/training/lora.mdx:118]: See if the current changes make sense.
